File: src/cpp/device_detection.cpp
#include "device_detection.h"

namespace fastly::device_detection {

std::optional<Device> lookup(std::string user_agent) {
  auto ptr{
      fastly::sys::device_detection::f_device_detection_lookup(user_agent)};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<Device>(
        Device(::rust::Box<fastly::sys::device_detection::Device>::from_raw(
            std::move(ptr))));
  }
}

std::optional<std::string> Device::device_name() {
  std::string out;
  auto some{this->dev->device_name(out)};
  if (!some) {
    return std::nullopt;
  } else {
    return std::optional<std::string>{out};
  }
}

std::optional<std::string> Device::brand() {
  std::string out;
  auto some{this->dev->brand(out)};
  if (!some) {
    return std::nullopt;
  } else {
    return std::optional<std::string>{out};
  }
}

std::optional<std::string> Device::model() {
  std::string out;
  auto some{this->dev->model(out)};
  if (!some) {
    return std::nullopt;
  } else {
    return std::optional<std::string>{out};
  }
}

std::optional<std::string> Device::hwtype() {
  std::string out;
  auto some{this->dev->hwtype(out)};
  if (!some) {
    return std::nullopt;
  } else {
    return std::optional<std::string>{out};
  }
}

std::optional<bool> Device::is_ereader() {
  auto ptr{this->dev->is_ereader()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_gameconsole() {
  auto ptr{this->dev->is_gameconsole()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_mediaplayer() {
  auto ptr{this->dev->is_mediaplayer()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_mobile() {
  auto ptr{this->dev->is_mobile()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_smarttv() {
  auto ptr{this->dev->is_smarttv()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_tablet() {
  auto ptr{this->dev->is_tablet()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_tvplayer() {
  auto ptr{this->dev->is_tvplayer()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_desktop() {
  auto ptr{this->dev->is_desktop()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

std::optional<bool> Device::is_touchscreen() {
  auto ptr{this->dev->is_touchscreen()};
  if (ptr == nullptr) {
    return std::nullopt;
  } else {
    return std::optional<bool>(*ptr);
  }
}

} // namespace fastly::device_detection

File: src/device_detection.rs
use std::pin::Pin;

use cxx::CxxString;

pub struct Device(pub(crate) fastly::device_detection::Device);

// Just used to force generation of symbols, since we're otherwise doing raw pointers.
pub fn f_device_detection_noop(dev: Box<Device>) -> Box<Device> {
    dev
}

pub fn f_device_detection_lookup(user_agent: &CxxString) -> *mut Device {
    if let Some(dev) = fastly::device_detection::lookup(user_agent.to_string_lossy().as_ref()) {
        Box::into_raw(Box::new(Device(dev)))
    } else {
        std::ptr::null_mut()
    }
}

impl Device {
    pub fn device_name(&self, out: Pin<&mut CxxString>) -> bool {
        if let Some(b) = self.0.device_name() {
            out.push_str(b);
            true
        } else {
            false
        }
    }
    pub fn brand(&self, out: Pin<&mut CxxString>) -> bool {
        if let Some(b) = self.0.brand() {
            out.push_str(b);
            true
        } else {
            false
        }
    }
    pub fn model(&self, out: Pin<&mut CxxString>) -> bool {
        if let Some(b) = self.0.model() {
            out.push_str(b);
            true
        } else {
            false
        }
    }
    pub fn hwtype(&self, out: Pin<&mut CxxString>) -> bool {
        if let Some(b) = self.0.model() {
            out.push_str(b);
            true
        } else {
            false
        }
    }
    pub fn is_ereader(&self) -> *const bool {
        if let Some(b) = self.0.is_ereader() {
            Box::into_raw(Box::new(b))
        } else {
            std::ptr::null()
        }
    }
    pub fn is_gameconsole(&self) -> *const bool {
        if let Some(b) = self.0.is_gameconsole() {
            Box::into_raw(Box::new(b))
        } else {
            std::ptr::null()
        }
    }
    pub fn is_mediaplayer(&self) -> *const bool {
        if let Some(b) = self.0.is_mediaplayer() {
            Box::into_raw(Box::new(b))
        } else {
            std::ptr::null()
        }
    }
    pub fn is_mobile(&self) -> *const bool {
        if let Some(b) = self.0.is_mobile() {
            Box::into_raw(Box::new(b))
        } else {
            std::ptr::null()
        }
    }
    pub fn is_smarttv(&self) -> *const bool {
        if let Some(b) = self.0.is_smarttv() {
            Box::into_raw(Box::new(b))
        } else {
            std::ptr::null()
        }
    }
    pub fn is_tablet(&self) -> *const bool {
        if let Some(b) = self.0.is_tablet() {
            Box::into_raw(Box::new(b))
        } else {
            std::ptr::null()
        }
    }
    pub fn is_tvplayer(&self) -> *const bool {
        if let Some(b) = self.0.is_tvplayer() {
            Box::into_raw(Box::new(b))
        } else {
            std::ptr::null()
        }
    }
    pub fn is_desktop(&self) -> *const bool {
        if let Some(b) = self.0.is_desktop() {
            Box::into_raw(Box::new(b))
        } else {
            std::ptr::null()
        }
    }
    pub fn is_touchscreen(&self) -> *const bool {
        if let Some(b) = self.0.is_touchscreen() {
            Box::into_raw(Box::new(b))
        } else {
            std::ptr::null()
        }
    }
}
